components/net/http_loader.rs

```diff
@@ -1020,51 +1036,127 @@ fn http_network_or_cache_fetch(
     // Step 5.18
     // TODO If there’s a proxy-authentication entry, use it as appropriate.
 
-    // Step 5.19
-    if let Ok(http_cache) = context.state.http_cache.read() {
-        if let Some(response_from_cache) = http_cache.construct_response(&http_request, done_chan) {
-            let response_headers = response_from_cache.response.headers.clone();
-            // Substep 1, 2, 3, 4
-            let (cached_response, needs_revalidation) =
-                match (http_request.cache_mode, &http_request.mode) {
-                    (CacheMode::ForceCache, _) => (Some(response_from_cache.response), false),
-                    (CacheMode::OnlyIfCached, &RequestMode::SameOrigin) => {
-                        (Some(response_from_cache.response), false)
-                    },
-                    (CacheMode::OnlyIfCached, _) |
-                    (CacheMode::NoStore, _) |
-                    (CacheMode::Reload, _) => (None, false),
-                    (_, _) => (
-                        Some(response_from_cache.response),
-                        response_from_cache.needs_validation,
-                    ),
-                };
-            if needs_revalidation {
-                revalidating_flag = true;
-                // Substep 5
-                if let Some(http_date) = response_headers.typed_get::<LastModified>() {
-                    let http_date: SystemTime = http_date.into();
-                    http_request
-                        .headers
-                        .typed_insert(IfModifiedSince::from(http_date));
+    // If the cache is not ready to construct a response, wait.
+    //
+    // The cache is not ready if a previous fetch checked the cache, found nothing,
+    // and moved on to a network fetch, and hasn't updated the cache yet with a pending resource.
+    //
+    // Note that this is a different workflow from the one involving `wait_for_cached_response`.
+    // That one happens when a fetch gets a cache hit, and the resource is pending completion from the network.
+    {
+        let (lock, cvar) = {
+            let mut state_map = context.state.http_cache_state.lock().unwrap();
+            let entry_key = CacheKey::new(http_request.clone());
+            &*state_map
+                .entry(entry_key)
+                .or_insert(Arc::new((
+                    Mutex::new(HttpCacheEntryState::ReadyToConstruct),
+                    Condvar::new(),
+                )))
+                .clone()
+        };
+
+        // Start of critical section on http-cache state.
+        let mut state = lock.lock().unwrap();
+        while let HttpCacheEntryState::PendingStore(_) = *state {
+            let (current_state, time_out) = cvar
+                .wait_timeout(state, Duration::from_millis(500))
```

Comment:
Waiting while holding a lock?

Reply:
https://doc.servo.org/std/sync/struct.Condvar.html#method.wait_timeout
The lock is not held while waiting.